### PR TITLE
[SPARK-39760][PYTHON] Support Varchar in PySpark

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Support Varchar in PySpark

### Why are the changes needed?
function parity

### Does this PR introduce _any_ user-facing change?
yes, new datatype supported

### How was this patch tested?
1, added UT;
2, manually check against the scala side:

```python

In [1]: from pyspark.sql.types import *
   ...: from pyspark.sql.functions import *
   ...: 
   ...: df = spark.createDataFrame([(1,), (11,)], ["value"])
   ...: ret = df.select(col("value").cast(VarcharType(10))).collect()
   ...: 
22/07/13 17:17:07 WARN CharVarcharUtils: The Spark cast operator does not support char/varchar type and simply treats them as string type. Please use string type directly to avoid confusion. Otherwise, you can set spark.sql.legacy.charVarcharAsString to true, so that Spark treat them as string type as same as Spark 3.0 and earlier
                                                                                
In [2]: 

In [2]: schema = StructType([StructField("a", IntegerType(), True), (StructField("v", VarcharType(10), True))])
   ...: description = "this a table created via Catalog.createTable()"
   ...: table = spark.catalog.createTable("tab3_via_catalog", schema=schema, description=description)
   ...: table.schema
   ...: 
Out[2]: StructType([StructField('a', IntegerType(), True), StructField('v', StringType(), True)])
```

```scala
scala> import org.apache.spark.sql.types._
import org.apache.spark.sql.types._

scala> import org.apache.spark.sql.functions._
import org.apache.spark.sql.functions._

scala> val df = spark.range(0, 10).selectExpr(" id AS value")
df: org.apache.spark.sql.DataFrame = [value: bigint]

scala> val ret = df.select(col("value").cast(VarcharType(10))).collect()
22/07/13 17:28:56 WARN CharVarcharUtils: The Spark cast operator does not support char/varchar type and simply treats them as string type. Please use string type directly to avoid confusion. Otherwise, you can set spark.sql.legacy.charVarcharAsString to true, so that Spark treat them as string type as same as Spark 3.0 and earlier
ret: Array[org.apache.spark.sql.Row] = Array([0], [1], [2], [3], [4], [5], [6], [7], [8], [9])

scala> 

scala> val schema = StructType(StructField("a", IntegerType, true) :: (StructField("v", VarcharType(10), true) :: Nil))
schema: org.apache.spark.sql.types.StructType = StructType(StructField(a,IntegerType,true),StructField(v,VarcharType(10),true))

scala> val description = "this a table created via Catalog.createTable()"
description: String = this a table created via Catalog.createTable()

scala> val table = spark.catalog.createTable("tab3_via_catalog", source="json", schema=schema, description=description, options=Map.empty[String, String])
table: org.apache.spark.sql.DataFrame = [a: int, v: string]

scala> table.schema
res0: org.apache.spark.sql.types.StructType = StructType(StructField(a,IntegerType,true),StructField(v,StringType,true))
```

### CODE REVIEW

**Comment by zhengruifeng** [comment]
gentle ping @HyukjinKwon @cloud-fan

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -181,6 +182,29 @@ class StringType(AtomicType, metaclass=DataTypeSingleton):
     pass
 
 
+class VarcharType(AtomicType):
+    """Varchar data type
+
+    Parameters
+    ----------
+    length : int
+    the length limitation. Data writing will fail if the input
```
Comment (xinrong-meng) [inline python/pyspark/sql/types.py:191]: nit: indent?

File: python/pyspark/sql/types.py [hunk 2]
```diff
@@ -1659,8 +1698,8 @@ def verify_acceptable_types(obj: Any) -> None:
                 new_msg("%s can not accept object %r in type %s" % (dataType, obj, type(obj)))
             )
 
-    if isinstance(dataType, StringType):
-        # StringType can work with any types
+    if isinstance(dataType, StringType) or isinstance(dataType, VarcharType):
```
Comment (xinrong-meng) [inline python/pyspark/sql/types.py:1701]: nit: `isinstance(dataType, (StringType, VarcharType))`?

File: python/pyspark/sql/types.py [hunk 3]
```diff
@@ -181,6 +182,29 @@ class StringType(AtomicType, metaclass=DataTypeSingleton):
     pass
 
 
+class VarcharType(AtomicType):
```
Comment (xinrong-meng) [inline python/pyspark/sql/types.py:185]: `metaclass=DataTypeSingleton`?

File: python/pyspark/sql/tests/test_types.py [hunk 4]
```diff
@@ -1132,6 +1138,14 @@ def test_decimal_type(self):
         t3 = DecimalType(8)
         self.assertNotEqual(t2, t3)
 
+    def test_varchar_type(self):
+        v1 = VarcharType(10)
+        v2 = VarcharType(20)
+        self.assertTrue(v2 is not v1)
+        self.assertNotEqual(v1, v2)
+        v3 = VarcharType(10)
+        self.assertEqual(v1, v3)
```
Comment (xinrong-meng) [inline python/pyspark/sql/tests/test_types.py:1147]: Shall we check if `v1 is v3`?

File: python/pyspark/sql/types.py (hunk 3 above)
Comment (zhengruifeng) [reply python/pyspark/sql/types.py:185]: I firstly tried this, but it cause initialization error due to the `__call__` in `DataTypeSingleton`:
a type mixin with `DataTypeSingleton` should support constructor without parameters.

File: python/pyspark/sql/types.py (hunk 2 above)
Comment (zhengruifeng) [reply python/pyspark/sql/types.py:1701]: nice!

File: python/pyspark/sql/tests/test_types.py (hunk 4 above)
Comment (xinrong-meng) [reply python/pyspark/sql/tests/test_types.py:1147]: `v1 is v3` should be True after parametric singleton is introduced. I will adjust that together in the followup.

**Comment by xinrong-meng** [comment]
LGTM! Thanks!

**Review by xinrong-meng — APPROVED** [review]

File: python/pyspark/sql/types.py (hunk 3 above)
Comment (xinrong-meng) [reply python/pyspark/sql/types.py:185]: https://issues.apache.org/jira/browse/SPARK-39794

**Review by HyukjinKwon — APPROVED** [review]
We'll have to fix the codes when `spark.sql.execution.arrow.pyspark.enabled` is enabled, and Py4J + Python UDFs too. But we can do it separately. See also https://issues.apache.org/jira/browse/SPARK-37275

**Comment by zhengruifeng** [comment]
Merged to master, thank you all!

@HyukjinKwon I will send followup PR for arrow/py4j/udf, thanks for pointing out it!

**Comment by cloud-fan** [comment]
late LGTM, do we want to support `CharType` in pyspark?

**Comment by zhengruifeng** [comment]
@cloud-fan  I think so, let me add it in near future.

@HyukjinKwon Btw, I guess we may not need to add extra support for pyarrow or py4j+python UDF, because it seems that there is not a class for char/varchar instances in scala or python (pandas/numpy/built-in), and they are treated as string internally.

There is also a warning message if we want to cast to char/varchar, in `CharVarcharUtils`
`The Spark cast operator does not support char/varchar type and simply treats them as string type. Please use string type directly to avoid confusion.`

**Pull request closed without merging**